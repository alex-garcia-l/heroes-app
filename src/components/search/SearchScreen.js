import React, { useMemo } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import { getHerosByName } from '../../selectors/hero/getHerosByName';
import { HeroCard } from '../heros/HeroCard';

export const SearchScreen = () => {

  const navigate = useNavigate();
  const location = useLocation();
  const query = queryString.parse(location.search);
  const { q = '' } = query;

  const initialValues = {
    inputSearch: q
  };

  const [{ inputSearch }, handleInputChange] = useForm(initialValues);
  const heroesFiltered = useMemo(() => getHerosByName(q), [q]);

  const handleSubmit = (evt) => {
    evt.preventDefault();
    navigate(`?q=${inputSearch}`);
  }
  return (
    <div className="row">
      <div className="col-12">
        <h2>Search</h2>
        <hr />
      </div>

      <div className="col-12 col-sm-6 m-auto my-5">
        <form className="" onSubmit={handleSubmit}>
          <div className="input-group">
            <input
              type="text"
              name="inputSearch"
              className="form-control"
              placeholder="Search a hero"
              autoComplete="off"
              value={inputSearch}
              onChange={handleInputChange}
            />
            <button
              type="submit"
              className="btn btn-outline-primary"
            >
              Search
            </button>
          </div>
        </form>
      </div>


      <div className="col-12">
        <h2>Heros</h2>
        <hr />

        {
          q === ''
            ? <div className="alert alert-info">Search a hero</div>
            : (heroesFiltered.length === 0)
              && <div className="alert alert-danger">Not hero to match: <span className="fw-bold">{q}</span></div>
        }

        <div className="row">
          {
            heroesFiltered.map(hero => (
              <HeroCard
                key={hero.id}
                {...hero}
              />
            ))
          }
        </div>
      </div>
    </div>
  );
};
