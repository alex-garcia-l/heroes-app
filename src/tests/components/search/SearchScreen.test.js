import { mount } from 'enzyme';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

// mook debe de ir
// evaluar cualquier hook
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate
}));

describe('Pruebas en el componente <SearchScreen />', () => {

  test('debe mostrarse correctamente con valores por defecto', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/home']}>
        <Routes>
          <Route path="home" element={<SearchScreen />} />
        </Routes>
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
  });

  test('debe mostrar a batman y el input con el valor al queryString', () => {
    const query = 'batman';
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/home?q=${query}`]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toBe(query);
    expect(wrapper.find('.card').length).toBeGreaterThanOrEqual(1);
  });

  test('debe de mostrar un error si no se encuntra el hero', () => {
    const query = 'batman123';
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/home?q=${query}`]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-danger').exists()).toBe(true);
    expect(wrapper.find('.alert-danger').text().trim()).toBe(`Not hero to match: ${query}`);
  });

  test('debe de llamar la funciÃ³n handleSubmit', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/home']}>
        <SearchScreen />
      </MemoryRouter>
    );

    const value = 'batman';

    wrapper.find('input').simulate('change', {
      target: {
        name: 'inputSearch',
        value: value
      }
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault() { }
    });

    expect(mockNavigate).toHaveBeenCalled();
    expect(mockNavigate).toHaveBeenCalledWith(`?q=${value}`);
  });

});
