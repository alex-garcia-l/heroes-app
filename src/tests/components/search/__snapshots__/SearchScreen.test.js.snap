// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <SearchScreen /> debe de mostrar un error si no se encuntra el hero 1`] = `
<div
  className="row"
>
  <div
    className="col-12"
  >
    <h2>
      Search
    </h2>
    <hr />
  </div>
  <div
    className="col-12 col-sm-6 m-auto my-5"
  >
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="input-group"
      >
        <input
          autoComplete="off"
          className="form-control"
          name="inputSearch"
          onChange={[Function]}
          placeholder="Search a hero"
          type="text"
          value="batman123"
        />
        <button
          className="btn btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <div
    className="col-12"
  >
    <h2>
      Heros
    </h2>
    <hr />
    <div
      className="alert alert-danger"
    >
      Not hero to match: 
      <span
        className="fw-bold"
      >
        batman123
      </span>
    </div>
    <div
      className="row"
    />
  </div>
</div>
`;

exports[`Pruebas en el componente <SearchScreen /> debe mostrar a batman y el input con el valor al queryString 1`] = `
<div
  className="row"
>
  <div
    className="col-12"
  >
    <h2>
      Search
    </h2>
    <hr />
  </div>
  <div
    className="col-12 col-sm-6 m-auto my-5"
  >
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="input-group"
      >
        <input
          autoComplete="off"
          className="form-control"
          name="inputSearch"
          onChange={[Function]}
          placeholder="Search a hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <div
    className="col-12"
  >
    <h2>
      Heros
    </h2>
    <hr />
    <div
      className="row"
    >
      <div
        className="col-12 col-sm-6 col-lg-4 animate__animated animate__fadeIn"
      >
        <div
          className="card mb-3"
        >
          <div
            className="row g-0"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="img-fluid rounded-start"
                src="/images/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small>
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Ver m√°s...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente <SearchScreen /> debe mostrarse correctamente con valores por defecto 1`] = `
<div
  className="row"
>
  <div
    className="col-12"
  >
    <h2>
      Search
    </h2>
    <hr />
  </div>
  <div
    className="col-12 col-sm-6 m-auto my-5"
  >
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="input-group"
      >
        <input
          autoComplete="off"
          className="form-control"
          name="inputSearch"
          onChange={[Function]}
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <div
    className="col-12"
  >
    <h2>
      Heros
    </h2>
    <hr />
    <div
      className="alert alert-info"
    >
      Search a hero
    </div>
    <div
      className="row"
    />
  </div>
</div>
`;
